File: backend/src/main/java/fr/agroscan/service/AdminUserService.java
package fr.agroscan.service;

import fr.agroscan.domain.AppUser;
import fr.agroscan.repository.AdminUserSummary;
import fr.agroscan.repository.AppUserRepository;
import fr.agroscan.repository.RoleRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Locale;
import java.util.Map;

@Service
public class AdminUserService {

    private static final Map<String, String> SORT_FIELDS = Map.of(
            "firstName", "firstName",
            "lastName", "lastName",
            "email", "email",
            "role", "role.name",
            "createdAt", "createdAt"
    );

    private final AppUserRepository userRepository;
    private final RoleRepository roleRepository;
    private final TokenService tokenService;

    public AdminUserService(AppUserRepository userRepository, RoleRepository roleRepository, TokenService tokenService) {
        this.userRepository = userRepository;
        this.roleRepository = roleRepository;
        this.tokenService = tokenService;
    }

    @Transactional(readOnly = true)
    public Page<AdminUserSummary> search(String search, int page, int size, String sortBy, boolean ascending) {
        String field = SORT_FIELDS.getOrDefault(sortBy, "createdAt");
        Sort sort = Sort.by(ascending ? Sort.Direction.ASC : Sort.Direction.DESC, field);
        return userRepository.searchForAdmin(search == null ? "" : search.trim(), PageRequest.of(page, Math.min(size, 50), sort));
    }

    @Transactional
    public AppUser update(
            String adminEmail,
            Long userId,
            String firstName,
            String lastName,
            String email,
            String roleName,
            boolean enabled
    ) {
        AppUser user = requireUser(userId);
        String normalizedEmail = email.trim().toLowerCase(Locale.ROOT);
        if (user.getEmail().equalsIgnoreCase(adminEmail)) {
            throw new AdminOperationException("Modifiez votre propre compte depuis la page Mon profil");
        }
        if (!user.getEmail().equalsIgnoreCase(normalizedEmail) && userRepository.existsByEmailIgnoreCase(normalizedEmail)) {
            throw new EmailAlreadyUsedException();
        }
        if ("ADMIN".equals(user.getRole().getName()) && !"ADMIN".equals(roleName)
                && userRepository.countByRoleName("ADMIN") <= 1) {
            throw new AdminOperationException("Le dernier administrateur ne peut pas perdre son rôle");
        }
        var role = roleRepository.findByName(roleName)
                .orElseThrow(() -> new AdminOperationException("Rôle utilisateur invalide"));
        user.updateByAdmin(firstName.trim(), lastName.trim(), normalizedEmail, role, enabled);
        tokenService.revokeAllForUser(user.getId());
        return user;
    }

    @Transactional
    public void delete(String adminEmail, Long userId) {
        AppUser user = requireUser(userId);
        if (user.getEmail().equalsIgnoreCase(adminEmail)) {
            throw new AdminOperationException("Vous ne pouvez pas supprimer votre propre compte");
        }
        if ("ADMIN".equals(user.getRole().getName()) && userRepository.countByRoleName("ADMIN") <= 1) {
            throw new AdminOperationException("Le dernier administrateur ne peut pas être supprimé");
        }
        userRepository.delete(user);
    }

    private AppUser requireUser(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("Utilisateur introuvable"));
    }
}

File: frontend/src/app/core/admin-user.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { PageResponse } from './scan.service';

export type UserRole = 'USER' | 'ADMIN';

export interface AdminUser {
  id: number;
  email: string;
  firstName: string;
  lastName: string;
  role: UserRole;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface AdminUserSearch {
  search: string;
  page: number;
  size: number;
  sortBy: 'firstName' | 'lastName' | 'email' | 'role' | 'createdAt';
  ascending: boolean;
}

export interface UpdateAdminUser {
  email: string;
  firstName: string;
  lastName: string;
  role: UserRole;
  enabled: boolean;
}

@Injectable({ providedIn: 'root' })
export class AdminUserService {
  private readonly http = inject(HttpClient);

  search(search: AdminUserSearch): Observable<PageResponse<AdminUser>> {
    const params = new HttpParams()
      .set('search', search.search)
      .set('page', search.page)
      .set('size', search.size)
      .set('sortBy', search.sortBy)
      .set('ascending', search.ascending);
    return this.http.get<PageResponse<AdminUser>>('/api/admin/users', { params });
  }

  update(id: number, user: UpdateAdminUser): Observable<AdminUser> {
    return this.http.patch<AdminUser>(`/api/admin/users/${id}`, user);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`/api/admin/users/${id}`);
  }
}
